coderd/prometheusmetrics/aggregator.go

```diff
@@ -174,9 +233,66 @@ func NewMetricsAggregator(logger slog.Logger, registerer prometheus.Registerer,
 		storeSizeGauge:   storeSizeGauge,
 		updateHistogram:  updateHistogram,
 		cleanupHistogram: cleanupHistogram,
+
+		aggregateByLabels: aggregateByLabels,
 	}, nil
 }
 
+// labelAggregator is used to control cardinality of collected Prometheus metrics by pre-aggregating series based on given labels.
+type labelAggregator struct {
+	aggregations map[string]float64
+	metrics      map[string]annotatedMetric
+}
+
+func newLabelAggregator(size int) *labelAggregator {
+	return &labelAggregator{
+		aggregations: make(map[string]float64, size),
+		metrics:      make(map[string]annotatedMetric, size),
+	}
+}
+
+func (a *labelAggregator) aggregate(am annotatedMetric, labels []string) error {
```

Comment:
QQ: What is the extra overhead if `labelAggregator` is used together with default set of labels?

I'm also curious if this can be implemented using a strategy pattern, aggregated vs. non-aggregated. It is not a mandatory requirement, rather a challenge. I'm afraid that messing up with the current aggregator may result in unexpected events/errors.

Reply:
> QQ: What is the extra overhead if labelAggregator is used together with default set of labels?

```go
goos: darwin
goarch: arm64
pkg: github.com/coder/coder/v2/coderd/prometheusmetrics
                          │ /tmp/no-aggregation.out │         /tmp/aggregation.out         │
                          │         sec/op          │    sec/op     vs base                │
_MetricsAggregator_Run-12              2.219m ± 34%   3.124m ± 31%  +40.77% (p=0.011 n=10)

                          │ /tmp/no-aggregation.out │         /tmp/aggregation.out         │
                          │          B/op           │     B/op      vs base                │
_MetricsAggregator_Run-12              2.800Mi ± 1%   3.288Mi ± 4%  +17.44% (p=0.000 n=10)

                          │ /tmp/no-aggregation.out │        /tmp/aggregation.out         │
                          │        allocs/op        │  allocs/op   vs base                │
_MetricsAggregator_Run-12               36.89k ± 0%   43.17k ± 0%  +17.03% (p=0.000 n=10)
```

Seems like quite a lot of overhead because of all the extra allocations: the profile indicates about 30% of time spent on GC, which may not be realistic in a normal running process but it's still indicative. I don't think it's too much of an issue unless we're aggregating metrics from hundreds or thousands of agents. What do you think?

> I'm also curious if this can be implemented using a strategy pattern, aggregated vs. non-aggregated. It is not a mandatory requirement, rather a challenge.

It'd certainly make it cleaner, although I doubt it's worth the time investment. What do you think?

> I'm afraid that messing up with the current aggregator may result in unexpected events/errors.

My changes don't alter the behaviour of the current aggregator since we're taking a copy of the `annotationMetric`s.

Reply:
Oh, I didn't expect the deep investigation :)

> It'd certainly make it cleaner, although I doubt it's worth the time investment. What do you think?

This is a question about our confidence in this change. If it seems to be stable/not loosing metrics, I guess we can risk and leave it as the only/default implementation 👍 

Anyway, feel free to resolve this comment. Thanks!